lib/aiken/rationals.ak

```diff
@@ -0,0 +1,628 @@
+}
+
+test mul_1() {
+  ( unsafe_ratio(2, 3) |> mul(unsafe_ratio(3, 4)) ) == Rational {
+    numerator: 6,
+    denominator: 12,
+  }
+}
+
+test mul_2() {
+  ( unsafe_ratio(-2, 3) |> mul(unsafe_ratio(-3, 4)) ) == Rational {
+    numerator: 6,
+    denominator: 12,
+  }
+}
+
+test mul_3() {
+  (
+    unsafe_ratio(2, 5)
+      |> mul(unsafe_ratio(1, 8))
+      |> mul(unsafe_ratio(3, 10))
+      |> mul(unsafe_ratio(21, 100))
+      |> mul(unsafe_ratio(3, 5))
+      |> mul(unsafe_ratio(2, 8))
+      |> mul(unsafe_ratio(4, 10))
+      |> mul(unsafe_ratio(22, 100))
+      |> reduce()
+  ) == Rational { numerator: 2079, denominator: 50000000 }
+}
+
+/// Division
+/// quotient of two rational values
+/// none if the second value is 0
+/// ```aiken
+/// div(Rational { numerator: x, denominator: y }, Rational { numerator: a, denominator: b }) == Rational { numerator: x*b, denominator: y*a }
+/// ```
+pub fn div(a: Rational, b: Rational) -> Option<Rational> {
+  recip(b) |> option.map(fn(x) { x |> mul(a) })
+}
+
+/// Unsafe division
+/// quotient of two rational values
+/// panics if the second value is 0 (unsafe)
+/// ```aiken
+/// unsafe_ratio(x, y) |> unsafe_div(unsafe_ratio(a, b)) == Rational { numerator: x*b, denominator: y*a }
+/// ```
+pub fn unsafe_div(a: Rational, b: Rational) -> Rational {
+  unsafe_recip(b) |> mul(a)
+}
+
+test div_1() {
+  ( unsafe_ratio(2, 3) |> unsafe_div(unsafe_ratio(3, 4)) ) == Rational {
+    numerator: 8,
+    denominator: 9,
+  }
+}
+
+test div_2() {
+  ( unsafe_ratio(2, 3) |> unsafe_div(unsafe_ratio(-3, 4)) ) == Rational {
+    numerator: -8,
+    denominator: 9,
+  }
+}
+
+/// Ratio of integers, a rational
+/// none if the second value is 0
+/// ```aiken
+/// ratio(x, 0) == None
+/// ratio(x, y) == Some(Rational { numerator: x, denominator: y })
+/// ```
+pub fn ratio(numerator: Int, denominator: Int) -> Option<Rational> {
+  if denominator == 0 {
+    None
+  } else if denominator < 0 {
+    Some(Rational { numerator: -numerator, denominator: -denominator })
+  } else {
+    Some(Rational { numerator, denominator })
+  }
+}
+
+test ratio_zero() {
+  ratio(2, 0) == None
+}
+
+test ratio_1() {
+  ratio(2, 3) == Some(Rational { numerator: 2, denominator: 3 })
+}
+
+test ratio_2() {
+  ratio(-2, 3) == Some(Rational { numerator: -2, denominator: 3 })
+}
+
+test ratio_3() {
+  ratio(2, -3) == Some(Rational { numerator: -2, denominator: 3 })
+}
+
+test ratio_4() {
+  ratio(2, 4) == Some(Rational { numerator: 2, denominator: 4 })
+}
+
+test ratio_5() {
+  ratio(-2, -3) == Some(Rational { numerator: 2, denominator: 3 })
+}
+
+test ratio_6() {
+  ratio(-2, -4) == Some(Rational { numerator: 2, denominator: 4 })
+}
+
+/// Create a new Rational
+/// quotient of two integers
+/// panics if the second value is 0 (unsafe)
+/// ```aiken
+/// unsafe_ratio(x, y) == Rational { numerator: x, denominator: y }
+/// unsafe_ratio(x, 0) == error
+/// ```
+pub fn unsafe_ratio(numerator: Int, denominator: Int) -> Rational {
```

Comment:
```suggestion
fn unsafe_ratio(numerator: Int, denominator: Int) -> Rational {
```

Reply:
This is the standard library, and it defines an abstraction. If there are backdoors to bypass the abstraction then the abstraction is arguably pointless. 

Hence if we intend to introduce a `Rational` module to the standard library, I expect it to have a non leaky abstraction. If by "better" you also mean more performant, I'd like to see benchmarks to justify this claim as I hardly believe this to be sufficient to justify introducing possibly hard to debug catastrophic errors.

Reply:
> If there are backdoors to bypass the abstraction then the abstraction is arguably pointless.

It is not bypassing the abstraction. Both the invariants are still enforced: negative sign in the numerator and no zero allowed in the denominator. It is not possible to use `unsafe_ratio` to create an invalid `Rational`.

> If by "better" you also mean more performant, ...

It is better in the sense of there being use cases where `Option<Rational>` is unnecessary. Having to deal with `Option<a>` creates more overhead than just using `a`. Even if the over head is slight, smart contracts don't have that much leeway in the available script size and execution units.

> ...introducing possibly hard to debug catastrophic errors.

I don't think `unsafe_ratio` introduces any catastrophic errors at all since there is an appropriate error message that is thrown in the only possible scenario where it can fail. Any bug due to `unsafe_ratio` being used with a zero denominator is easy to track down, especially if developers only use `unsafe_ratio` in appropriate scenarios.

Also, it must go in the same module where `Rational` is defined because there is no way to define `unsafe_ratio` without the constructors. Therefore, if a developer did want to use it, they would need to create their own `Rational` type and interface just to have this function. This interface would then be incompatible with the `Rational` in the standard library.

Reply:
I think I understand. There's no abstraction being broken here, this is just a version of it that can fail. I think this kind of stuff will be less confusing/adhoc once we introduce into the type system the ability to declare that a function body must be pure and can't call expect or anything else that fails. This is also nice because there are some builtins that can fail and that it's obvious atm by just looking at the builtins signature.

Reply:
https://github.com/aiken-lang/stdlib/pull/47#discussion_r1231897097